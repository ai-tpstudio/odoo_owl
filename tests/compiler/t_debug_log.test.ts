import { renderToString, snapshotTemplate } from "../helpers";

// -----------------------------------------------------------------------------
// debugging
// -----------------------------------------------------------------------------

describe("debugging", () => {
  test("t-debug", () => {
    const consoleLog = console.log;
    console.log = jest.fn();
    const template = `<div t-debug=""><t t-if="true"><span t-debug="">hey</span></t></div>`;
    snapshotTemplate(template);
    expect(console.log).toHaveBeenCalledTimes(1);
    console.log = consoleLog;
  });

  test("t-debug on sub template", () => {
    const consoleLog = console.log;
    console.log = jest.fn();
    let template = `<p t-debug="">coucou</p>`;
    snapshotTemplate(template);
    template = `<div><t t-call="sub"/></div>`;
    snapshotTemplate(template);
    expect(console.log).toHaveBeenCalledTimes(1);
    console.log = consoleLog;
  });

  test("t-log", () => {
    const consoleLog = console.log;
    console.log = jest.fn();

    const template = `<div>
            <t t-set="foo" t-value="42"/>
            <t t-log="foo + 3"/>
          </div>`;
    snapshotTemplate(template);
    renderToString(template);
    expect(console.log).toHaveBeenCalledWith(45);
    console.log = consoleLog;
  });

  test("t-log: interaction with t-set", () => {
    const consoleLog = console.log;
    console.log = jest.fn();

    const template = `
      <t>
        <t t-log="foo" t-set="foo" t-value="42"/>
        <t t-log="bar" t-set="bar" t-value="49"/>
        <span t-esc="foo + bar"/>
      </t>
    `;
    snapshotTemplate(template);
    renderToString(template);
    expect(console.log).toHaveBeenCalledWith(undefined);
    expect(console.log).toHaveBeenCalledWith(undefined);
    console.log = consoleLog;
  });

  test("t-debug: interaction with t-set", () => {
    const template = `
      <t>
        <t t-debug="" t-set="foo" t-value="42"/>
        <t t-debug="" t-set="bar" t-value="49"/>
        <span t-esc="foo + bar"/>
      </t>
    `;
    snapshotTemplate(template);
    renderToString(template);
  });
});
