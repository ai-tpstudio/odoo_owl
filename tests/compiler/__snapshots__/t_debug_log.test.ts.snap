// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debugging t-debug 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  let block2 = createBlock(\`<span>hey</span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    debugger;
    let b2;
    if (true) {
      debugger;
      b2 = block2();
    }
    return block1([], [b2]);
  }
}"
`;

exports[`debugging t-debug on sub template 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<p>coucou</p>\`);
  
  return function template(ctx, node, key = \\"\\") {
    debugger;
    return block1();
  }
}"
`;

exports[`debugging t-debug on sub template 2`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  const callTemplate_1 = app.getTemplate(\`sub\`);
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    const b2 = callTemplate_1.call(this, ctx, node, key + \`__1\`);
    return block1([], [b2]);
  }
}"
`;

exports[`debugging t-debug: interaction with t-set 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { isBoundary, withDefault, setContextValue } = helpers;
  
  let block1 = createBlock(\`<span><block-text-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    ctx[isBoundary] = 1
    debugger;
    setContextValue(ctx, \\"foo\\", 42);
    debugger;
    setContextValue(ctx, \\"bar\\", 49);
    let txt1 = ctx['foo']+ctx['bar'];
    return block1([txt1]);
  }
}"
`;

exports[`debugging t-log 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { isBoundary, withDefault, setContextValue } = helpers;
  
  let block1 = createBlock(\`<div/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    ctx[isBoundary] = 1
    setContextValue(ctx, \\"foo\\", 42);
    console.log(ctx['foo']+3);
    return block1();
  }
}"
`;

exports[`debugging t-log: interaction with t-set 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { isBoundary, withDefault, setContextValue } = helpers;
  
  let block1 = createBlock(\`<span><block-text-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    ctx[isBoundary] = 1
    console.log(ctx['foo']);
    setContextValue(ctx, \\"foo\\", 42);
    console.log(ctx['bar']);
    setContextValue(ctx, \\"bar\\", 49);
    let txt1 = ctx['foo']+ctx['bar'];
    return block1([txt1]);
  }
}"
`;
